tests/acceptance/test_uboot_automation.py

```diff
@@ -0,0 +1,168 @@
+
+        meta_mender_uboot_rev = subprocess.check_output("git rev-list -n1 HEAD -- ../../meta-mender-core/recipes-bsp/u-boot",
+                                                        shell=True).strip()
+        if expected_meta_mender_uboot_rev != meta_mender_uboot_rev:
+            return False
+
+        return True
+
+    # No need to test this on non-vexpress-qemu. It is a very resource consuming
+    # test, and it is identical on all boards, since it internally tests all
+    # boards.
+    @pytest.mark.only_for_machine('vexpress-qemu')
+    @pytest.mark.min_mender_version('1.0.0')
+    def test_uboot_compile(self, bitbake_path):
+        """Test that our automatic patching of U-Boot still successfully builds
+        the expected number of boards."""
+
+        # THIS IS A SLOW RUNNING TEST!
+        # Use the revision inside the function below to optimize and skip
+        # checking based on version.
+        if self.revision_already_checked():
+            pytest.skip(reason="Revision of poky and u-boot already checked")
+
+        for task in ["do_provide_mender_defines", "prepare_recipe_sysroot"]:
+            subprocess.check_call("cd %s && bitbake -c %s u-boot" % (os.environ['BUILDDIR'], task),
+                                  shell=True)
+        bitbake_variables = get_bitbake_variables("u-boot")
+
+        env = copy.copy(os.environ)
+        env['UBOOT_SRC'] = bitbake_variables['S']
+        env['TESTS_DIR'] = os.getcwd()
+        env['LOGS'] = os.path.join(os.getcwd(), "test_uboot_compile-logs")
+        sanitized_makeflags = bitbake_variables['EXTRA_OEMAKE']
+        sanitized_makeflags = sanitized_makeflags.replace("\\\"", "\"")
+        sanitized_makeflags = re.sub(" +", " ", sanitized_makeflags)
+        # Compile all boards. The reason for using a makefile is to get easy
+        # parallelization.
+        subprocess.check_call("make -j %d -f %s %s" % (multiprocessing.cpu_count() + 1,
+                                                       os.path.join(env['TESTS_DIR'],
+                                                                    "files/Makefile.test_uboot_automation"),
+                                                       sanitized_makeflags),
+                              shell=True,
+                              env=env,
+                              stderr=subprocess.STDOUT)
+
+        # Now check that the ratio of compiled boards is as expected. This
+        # number may change over time as U-Boot changes, but big discrepancies
+        # should be checked out.
+        failed = 0.0
+        total = 0.0
+        for file in os.listdir(env['LOGS']):
+            total += 1
+            with open(os.path.join(env['LOGS'], file)) as fd:
+                if "AutoPatchFailed\n" in fd.readlines():
+                    failed += 1
+
+        # PLEASE UPDATE the version you used to find this number if you update it.
+        # From version: v2017.09
+        measured_failed_ratio = 747.0 / 1176.0
```

Comment:
The one problem I can see here is that we will need to update this somehow from time to time. What if new version of u-boot will cause increase/decrease which is in range (and we are not updating the number here), and then the following one will be out of boundaries. I know this is rather hard problem to solve, but something to think of.

Reply:
Maybe we could not fail the test but just print the message? Can we have some soft failures implemented somehow?

Reply:
> What if new version of u-boot will cause increase/decrease which is in range (and we are not updating the number here)

That's why I want it to fail. A significant change in either direction may signal a bug: Either that U-Boot has changed and the tool is now misbehaving, or that we are succeeding in patching suspiciously many boards, which may signal that there is a bug in the test itself. Both cases should be checked out, and if it looks legitimate, updating the number should be done to account for the new status. It's a bit crappy to have a test like this that we have to keep updated, but I don't see how else to test a heuristic tool like this.

I can try to improve the comment a bit more.